File: packages/book/src/type_system/42_structural_type/java/Main.java
public class Main {
  public static void main(String[] args) {
    Employee employee = new Employee("木村", 30);
    PartTime partTime = new PartTime("佐藤", 25);

    employee = partTime; // 名前の互換性があるため、エラーになります
    System.out.println("Employee: " + employee.name + ", Age: " + employee.age);
    System.out.println("PartTime: " + partTime.name + ", Age: " + partTime.age);
  }
}

File: packages/book/src/type_system/42_structural_type/main.ts
interface Employee {
  name: string;
  age: number;
}

interface PartTime {
  name: string;
  age: number;
}

const modifyName = (employee: Employee, name: string): Employee => ({
  ...employee,
  name,
});

{
  let employee: Employee = { name: 'Alice', age: 30 };
  const partTime: PartTime = { name: 'Bob', age: 25 };
  employee = partTime; // 構造の互換性があるため、エラーになりません
  modifyName(employee, 'Charlie'); // 関数シグニチャではPerson型を期待していますが、Employee型を渡してもエラーになりません
  console.log(employee);
}

export type { Employee, PartTime };
